x-pack/plugins/apm/public/hooks/use_transaction_distribution_fetcher.ts

```diff
@@ -5,112 +5,130 @@
  * 2.0.
  */
 
-import { flatten, omit, isEmpty } from 'lodash';
-import { useHistory } from 'react-router-dom';
-import { useFetcher } from './use_fetcher';
-import { toQuery, fromQuery } from '../components/shared/Links/url_helpers';
-import { maybe } from '../../common/utils/maybe';
-import { APIReturnType } from '../services/rest/createCallApmApi';
-import { useUrlParams } from '../context/url_params_context/use_url_params';
-import { useApmServiceContext } from '../context/apm_service/use_apm_service_context';
+import { useRef, useState } from 'react';
+import type { Subscription } from 'rxjs';
+import {
+  IKibanaSearchRequest,
+  IKibanaSearchResponse,
+  isCompleteResponse,
+  isErrorResponse,
+} from '../../../../../src/plugins/data/public';
+import type {
+  HistogramItem,
+  SearchServiceParams,
+  SearchServiceValue,
+} from '../../common/search_strategies/correlations/types';
+import { useKibana } from '../../../../../src/plugins/kibana_react/public';
+import { ApmPluginStartDeps } from '../plugin';
 
-type APIResponse = APIReturnType<'GET /api/apm/services/{serviceName}/transactions/charts/distribution'>;
+interface RawResponse {
+  percentileThresholdValue?: number;
+  took: number;
+  values: SearchServiceValue[];
+  overallHistogram: HistogramItem[];
+  log: string[];
+  ccsWarning: boolean;
+}
+
+export function useTransactionDistributionFetcher(
+  params: Omit<SearchServiceParams, 'analyzeCorrelations'>
+) {
+  const {
+    services: { data },
+  } = useKibana<ApmPluginStartDeps>();
 
-const INITIAL_DATA = {
-  buckets: [] as APIResponse['buckets'],
-  noHits: true,
-  bucketSize: 0,
-};
+  const [error, setError] = useState<Error>();
+  const [isComplete, setIsComplete] = useState(false);
+  const [isRunning, setIsRunning] = useState(false);
+  const [loaded, setLoaded] = useState<number>(0);
 
-export function useTransactionDistributionFetcher({
-  transactionName,
-  kuery,
-  environment,
-}: {
-  transactionName: string;
-  kuery: string;
-  environment: string;
-}) {
-  const { serviceName, transactionType } = useApmServiceContext();
+  const [transactionDistribution, setTansactionDistribution] = useState<
+    RawResponse['overallHistogram']
+  >();
+  const [percentileThresholdValue, setPercentileThresholdValue] = useState<
+    RawResponse['percentileThresholdValue']
+  >();
```

Comment:
Why a state handler for each property? Afaict you can replace `setTansactionDistribution`, `setPercentileThresholdValue`, `setTimeTook` and `setTotal` with:

```ts
setResponse(response)
```

Then in the return handler you can have something like: 

```ts
{
  error,
  response,
  status: { isRunning, isComplete, progress},
  startFetch,
  cancelFetch,  
}
```

Reply:
See https://github.com/elastic/kibana/pull/107266#discussion_r689343111